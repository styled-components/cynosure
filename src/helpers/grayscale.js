// @flow

import parseColorString from '../internalHelpers/_parseColorString'
import rgbToHsl from '../internalHelpers/_rgbToHsl'
import hslToHex from '../internalHelpers/_hslToHex'
import hslToRgb from '../internalHelpers/_hslToRgb'

/**
 * Converts the color to a grayscale, but reducing its saturation to 0.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   background: grayscale('#CCCD64'),
 *   background: grayscale('rgba(204,205,100,0.7)'),
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   background: ${grayscale('#CCCD64')};
 *   background: ${grayscale('rgba(204,205,100,0.7)')};
 * `
 *
 * // CSS in JS Output
 * element {
 *   background: "#999";
 *   background: "rgba(153,153,153,0.7)";
 * }
 */
function grayscale(color: string): string {
  // parse color string to hsl
  const hslColor = rgbToHsl(parseColorString(color))
  const value = {
    ...hslColor,
    saturation: 0,
  }
  return value.alpha >= 1 || value.alpha === undefined
    ? hslToHex(value.hue, value.saturation, value.lightness)
    : `rgba(${hslToRgb(value.hue, value.saturation, value.lightness)},${value.alpha})`
}

export default grayscale
