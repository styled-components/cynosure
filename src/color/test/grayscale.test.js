// @flow
import grayscale from '../grayscale'

describe('grayscale', () => {
  it('should grayscale a reduced hex color', () => {
    expect(grayscale('#444')).toEqual('#444')
  })

  it('should grayscale a hex color', () => {
    expect(grayscale('#CCCD64')).toEqual('#999')
  })

  it('should grayscale an 8-digit hex color', () => {
    expect(grayscale('#6564CDB3')).toEqual('rgba(153,153,153,0.7)')
  })

  it('should grayscale an 4-digit hex color', () => {
    expect(grayscale('#0f08')).toEqual('rgba(128,128,128,0.53)')
  })

  it('should grayscale a color with opacity', () => {
    expect(grayscale('rgba(101,100,205,0.7)')).toEqual('rgba(153,153,153,0.7)')
  })

  it('should return transparent when passed transparent', () => {
    expect(grayscale('transparent')).toEqual('transparent')
  })
})
