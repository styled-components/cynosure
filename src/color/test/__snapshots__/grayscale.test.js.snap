// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grayscale should grayscale a 8-digit hex color 1`] = `"rgba(153,153,153,0.4980392156862745)"`;

exports[`grayscale should grayscale a color with opacity 1`] = `"rgba(153,153,153,0.7)"`;

exports[`grayscale should grayscale a hex color 1`] = `"#999"`;

exports[`grayscale should grayscale a reduced hex color 1`] = `"#444"`;
